File: py/add-two-numbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @return a ListNode
    def addTwoNumbers(self, l1, l2):
        if l1 == None:
        	return l2
        if l2 == None:
        	return l1
        p = l1
        q = l2
        head = p
        carry = 0
        while True:
        	val = p.val + q.val + carry
        	if val >= 10:
        		carry = 1
        		val -= 10
        	else:
        		carry = 0
        	p.val = val
        	if p.next == None or q.next == None:
        		break
        	p = p.next
        	q = q.next
        if p.next == None:
        	p.next = q.next
        while p.next != None:
        	p = p.next
        	val = p.val + carry
        	if val >= 10:
        		carry = 1
        		val -= 10
        	else:
        		carry = 0
        	p.val = val
        if carry == 1:
        	p.next = ListNode(1)
        return head

    def reverse(self,head):
    	p = head
    	q = p.next
    	while q != None:
    		n = q.next
    		q.next = p
    		p = q
    		q = n
    		if q == None:
    			head.next = None
    			return p
    	return head

s = Solution()
n1 = ListNode(2)
n2 = ListNode(4)
n3 = ListNode(3)

n4 = ListNode(5)
n5 = ListNode(6)
n6 = ListNode(4)

n1.next = n2
n2.next = n3

n4.next = n5
n5.next = n6

p = s.addTwoNumbers(n1, n4)
while p != None:
	print p.val
	p = p.next

File: c/add_two_numbers.c
/*
You are given two linked lists representing two non-negative numbers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.
Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8

https://leetcode.com/problems/add-two-numbers/
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

#include "stdio.h"
#include "stdlib.h"

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    int carry = 0;
    struct ListNode* p = l1;
    struct ListNode* q = l2;
    if (p == NULL) return q;
    if (q == NULL) return p;
    while (1) {
        int num = p->val + (q != NULL ? q->val : 0) + carry;
        if (num >= 10) {
            num -= 10;
            carry = 1;
        } else {
            carry = 0;
        }
        p->val = num;
        if (p->next == NULL) {
            if (q != NULL && q->next != NULL) {
                p->next = q->next;
                q = NULL;
            } else {
                break;
            }
        }
        p = p->next;
        q = q != NULL ? q->next : NULL;
    }
    if (carry != 0) {
        struct ListNode* node = malloc(sizeof(struct ListNode));
        node->val = carry;
        node->next = NULL;
        p->next = node;
    }
    return l1;
}

int main() {
    // test
    struct ListNode* l1 = malloc(sizeof(struct ListNode));
    struct ListNode* l2 = malloc(sizeof(struct ListNode));
    struct ListNode* l3 = malloc(sizeof(struct ListNode));
    struct ListNode* l4 = malloc(sizeof(struct ListNode));
    struct ListNode* l5 = malloc(sizeof(struct ListNode));
    struct ListNode* l6 = malloc(sizeof(struct ListNode));
    l1->val = 2;
    l1->next = l2;
    l2->val = 4;
    l2->next = l3;
    l3->val = 3;
    l3->next = NULL;

    l4->val = 5;
    l4->next = l5;
    l5->val = 2;
    l5->next = l6;
    l6->val = 4;
    l6->next = NULL;

    struct ListNode* result = addTwoNumbers(l3, l4);
    while (result != NULL) {
        printf("%d ", result->val);
        result = result->next;
    }
    return 0;
}